code/modules/admin/verbs/lua/lua_state.dm

```diff
@@ -0,0 +1,125 @@
+GLOBAL_VAR_INIT(IsLuaCall, FALSE)
+GLOBAL_PROTECT(IsLuaCall)
+
+GLOBAL_DATUM(lua_usr, /mob)
+GLOBAL_PROTECT(lua_usr)
+
+/datum/lua_state
+	var/name
+
+	/// The internal ID of the lua state stored in auxlua's global map
+	var/internal_id
+
+	/// A log of every return, yield, and error for each chunk execution and function call
+	var/list/log = list()
+
+	/// A list of all the variables in the state's environment
+	var/list/globals = list()
+
+	/// A list in which to store datums and lists instantiated in lua, ensuring that they don't get garbage collected
+	var/list/references = list()
+
+/datum/lua_state/vv_edit_var(var_name, var_value)
+	. = ..()
+	if(var_name == NAMEOF(src, internal_id))
+		return FALSE
+
+/datum/lua_state/CanProcCall(procname)
+	if(GLOB.IsLuaCall) //No calling auxlua hooks from lua code
+		return FALSE
+	return ..()
+
+/datum/lua_state/New(_name)
+	if(SSlua.initialized != TRUE)
+		qdel(src)
+		return
+	name = _name
+	internal_id = __lua_new_state()
+
+/datum/lua_state/proc/handle_result(result)
+
+	// If this is a sleep, we need to add it to the subsystem's list of sleeps to run in the next fire
+	if(result["status"] == "sleeping")
+		SSlua.sleeps += src
+		// We only want to add the result to the state log if there is any meaningful information
+		// (particularly chunk source code), otherwise we'll get an unnecessarily flooded log at best
+		// or an OOM error at worst
+		if(!result["chunk"])
+			return
+	log += list(result)
+	// We want to return the return value(s) of executed code
+	if(result["status"] == "finished" || result["status"] == "yielded")
+		if(!length(result["param"]))
+			return
+		// Is there a meaningful difference between returning multiple values and returning a single list?
+		else if(length(result["param"]) == 1)
+			return result["param"][1]
+		else
+			return result["param"]
```

Comment:
Explain what this is doing and where it is to me please, this feels like it could cause trouble where it's not clear if something is returning a table or multiple elements.

Reply:
If a lua function returns a single value, auxlua will convert it into a 1-element list. If a callback or signal handler calls said function, expecting not-a-list, there will probably be a runtime. If it needs to be clear, the untouched return value is in the log that the UI reads, and I can have the UI show how many return values the function returned.